From c145106dcc6230fc18477ed24e1ee60f3375d52a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Jonas=20=C3=85dahl?= <jadahl@gmail.com>
Date: Sat, 12 Sep 2026 21:28:55 +0200
Subject: [PATCH] renderer/native: Always mark onscreens as detached when
 rebuilding views

When rebuilding the views, the views may get freed, meaning if we don't
mark them as detached (unset the view, disconnect signals), they will
still react to e.g. gamma changes, or privacy screen changes etc, where
the handles will try to interact with the now potentially freed view.

For this to happen, the onscreen must be kept alive longer than the
view; this may happen when a hotplug happens while in power save mode,
which will queue a symbolic power save page flip. If something triggers
one of the signals an onscreen listens on before the power save page
flip frees the onscreen, we'll hit the crash.

The added test case triggered this by changing the CRTC gamma before
leaving power save mode.

An example backtrace:

    #0  clutter_stage_view_schedule_update ()
    ...
    #5  g_signal_emit ()
    #6  meta_crtc_kms_set_gamma_lut ()
    #7  meta_crtc_set_gamma_lut ()
    #8  set_gamma_lut ()
    #9  meta_monitor_mode_foreach_crtc ()
    #10 meta_monitor_set_gamma_lut ()
    #11 update_white_point ()
    #12 do_update ()
    #13 in meta_color_device_update ()
    #14 update_device_properties ()
    #15 in meta_color_manager_monitors_changed ()
    #16 ensure_monitor_color_devices ()
    #17 meta_monitor_manager_notify_monitors_changed ()
    ...
    #34 g_main_loop_run ()
    #35 meta_context_run_main_loop ()

Closes: https://gitlab.gnome.org/GNOME/mutter/-/work_items/5042
Part-of: <https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/5324>
(cherry picked from commit 8e3870f45792836291fee329ed897da2d6a07c54)
---
 src/backends/meta-monitor-manager-private.h |  1 +
 src/backends/meta-monitor-private.h         |  1 +
 src/backends/native/meta-renderer-native.c  | 11 ++---
 src/tests/native-kms-device.c               | 46 +++++++++++++++++++++
 4 files changed, 54 insertions(+), 5 deletions(-)

diff --git a/src/backends/meta-monitor-manager-private.h b/src/backends/meta-monitor-manager-private.h
index 2e7ff97e32..cfce38d376 100644
--- a/src/backends/meta-monitor-manager-private.h
+++ b/src/backends/meta-monitor-manager-private.h
@@ -287,6 +287,7 @@ void                meta_monitor_manager_get_screen_size   (MetaMonitorManager *
                                                             int                *width,
                                                             int                *height);
 
+META_EXPORT_TEST
 MetaPowerSave       meta_monitor_manager_get_power_save_mode (MetaMonitorManager *manager);
 
 META_EXPORT_TEST
diff --git a/src/backends/meta-monitor-private.h b/src/backends/meta-monitor-private.h
index fcfa941262..c668f3c394 100644
--- a/src/backends/meta-monitor-private.h
+++ b/src/backends/meta-monitor-private.h
@@ -302,6 +302,7 @@ gboolean meta_monitor_get_min_refresh_rate (MetaMonitor *monitor,
 META_EXPORT_TEST
 size_t meta_monitor_get_gamma_lut_size (MetaMonitor *monitor);
 
+META_EXPORT_TEST
 void meta_monitor_set_gamma_lut (MetaMonitor        *monitor,
                                  const MetaGammaLut *lut);
 
diff --git a/src/backends/native/meta-renderer-native.c b/src/backends/native/meta-renderer-native.c
index 66b0e1e008..802a2b5d91 100644
--- a/src/backends/native/meta-renderer-native.c
+++ b/src/backends/native/meta-renderer-native.c
@@ -1662,8 +1662,6 @@ maybe_detach_onscreens (MetaRenderer *renderer)
   GList *l;
 
   power_save_mode = meta_monitor_manager_get_power_save_mode (monitor_manager);
-  if (power_save_mode != META_POWER_SAVE_ON)
-    return;
 
   views = meta_renderer_get_views (renderer);
   for (l = views; l; l = l->next)
@@ -1676,9 +1674,12 @@ maybe_detach_onscreens (MetaRenderer *renderer)
 
       meta_onscreen_native_detach (META_ONSCREEN_NATIVE (onscreen));
 
-      renderer_native->detached_onscreens =
-        g_list_prepend (renderer_native->detached_onscreens,
-                        g_object_ref (onscreen));
+      if (power_save_mode == META_POWER_SAVE_ON)
+        {
+          renderer_native->detached_onscreens =
+            g_list_prepend (renderer_native->detached_onscreens,
+                            g_object_ref (onscreen));
+        }
     }
 }
 
diff --git a/src/tests/native-kms-device.c b/src/tests/native-kms-device.c
index f9cadf3f5b..e9bff80107 100644
--- a/src/tests/native-kms-device.c
+++ b/src/tests/native-kms-device.c
@@ -18,6 +18,9 @@
 
 #include "config.h"
 
+#include "backends/meta-crtc.h"
+#include "backends/meta-monitor-manager-private.h"
+#include "backends/meta-monitor-private.h"
 #include "backends/native/meta-backend-native-private.h"
 #include "backends/native/meta-device-pool.h"
 #include "backends/native/meta-input-thread.h"
@@ -34,7 +37,9 @@
 #include "backends/native/meta-seat-native.h"
 #include "backends/native/meta-thread-impl.h"
 #include "meta-test/meta-context-test.h"
+#include "meta/meta-backend.h"
 #include "tests/meta-kms-test-utils.h"
+#include "tests/meta-monitor-test-utils.h"
 #include "tests/meta-test-utils.h"
 
 static MetaContext *test_context;
@@ -406,6 +411,45 @@ meta_test_kms_device_power_save (void)
                     meta_kms_crtc_get_id (crtc));
 }
 
+static void
+meta_test_kms_device_power_save_gamma_rebuild (void)
+{
+  MetaBackend *backend = meta_context_get_backend (test_context);
+  MetaMonitorManager *monitor_manager =
+    meta_backend_get_monitor_manager (backend);
+  ClutterStage *stage = CLUTTER_STAGE (meta_backend_get_stage (backend));
+  GList *monitors;
+  MetaMonitor *monitor;
+  size_t gamma_lut_size;
+  g_autoptr (MetaGammaLut) lut = NULL;
+
+  monitors = meta_monitor_manager_get_monitors (monitor_manager);
+  g_assert_nonnull (monitors);
+  monitor = META_MONITOR (monitors->data);
+
+  gamma_lut_size = meta_monitor_get_gamma_lut_size (monitor);
+  g_assert_cmpint (gamma_lut_size, >, 0);
+
+  meta_wait_for_presented (test_context);
+
+  meta_monitor_manager_set_power_save_mode (monitor_manager,
+                                            META_POWER_SAVE_OFF);
+  g_assert_cmpint (meta_monitor_manager_get_power_save_mode (monitor_manager),
+                   ==,
+                   META_POWER_SAVE_OFF);
+
+  clutter_actor_queue_redraw (CLUTTER_ACTOR (stage));
+  meta_wait_for_update (test_context);
+
+  meta_fake_hotplug (test_context);
+
+  lut = meta_gamma_lut_new_sized (gamma_lut_size);
+  meta_monitor_set_gamma_lut (monitor, lut);
+
+  meta_monitor_manager_set_power_save_mode (monitor_manager,
+                                            META_POWER_SAVE_ON);
+}
+
 static void
 done_update_result_feedback (const MetaKmsFeedback *feedback,
                              gpointer               user_data)
@@ -724,6 +768,8 @@ init_tests (void)
                    meta_test_kms_device_mode_set);
   g_test_add_func ("/backends/native/kms/device/power-save",
                    meta_test_kms_device_power_save);
+  g_test_add_func ("/backends/native/kms/device/power-save-gamma-rebuild",
+                   meta_test_kms_device_power_save_gamma_rebuild);
   g_test_add_func ("/backends/native/kms/device/discard-disabled",
                    meta_test_kms_device_discard_disabled);
   g_test_add_func ("/backends/native/kms/device/empty-update",
-- 
2.53.0

